crates/sats/src/slim_slice.rs

```diff
@@ -0,0 +1,1060 @@
+        drop(into_box(this));
+    }
+}
+
+impl<T> SlimSliceBox<T> {
+    /// Converts `boxed` to `Self` without checking `boxed.len() <= u32::MAX`.
+    ///
+    /// Only safe to call if the constraint above is satisfied.
+    #[allow(clippy::boxed_local)]
+    // Clippy doesn't seem to consider unsafe code here.
+    unsafe fn from_boxed_unchecked(boxed: Box<[T]>) -> Self {
+        let len = boxed.len();
+        let ptr = Box::into_raw(boxed) as *mut T;
+        // SAFETY: `Box<T>`'s ptr was a `NonNull<T>` already.
+        let raw = SlimRawSlice::from_len_ptr(len, ptr);
+        let owned = PhantomData;
+        Self { raw, owned }
+    }
+
+    /// Converts `boxed: Box<[T]>` into `SlimSlice<T>`.
+    ///
+    /// Panics when `boxed.len() > u32::MAX`.
+    pub fn from_boxed(boxed: Box<[T]>) -> Self {
+        expect_fit(boxed)
+    }
+
+    /// Converts `vec: Vec<T>` into `SlimSlice<T>`.
+    ///
+    /// Panics when `vec.len() > u32::MAX`.
+    pub fn from_vec(vec: Vec<T>) -> Self {
+        Self::from_boxed(vec.into())
+    }
+
+    /// Returns a limited shared slice to this boxed slice.
+    #[allow(clippy::needless_lifetimes)]
+    pub fn shared_ref<'a>(&'a self) -> &'a SlimSlice<'a, T> {
+        // SAFETY: The reference lives as long as `self`.
+        // By virtue of `repr(transparent)` we're also allowed these reference casts.
+        unsafe { mem::transmute(self) }
+    }
+
+    /// Returns a limited mutable slice to this boxed slice.
+    #[allow(clippy::needless_lifetimes)]
+    pub fn exclusive_ref<'a>(&'a mut self) -> &'a mut SlimSliceMut<'a, T> {
+        // SAFETY: The reference lives as long as `self`
+        // and we have exclusive access to the heap data thanks to `&'a mut self`.
+        // By virtue of `repr(transparent)` we're also allowed these reference casts.
+        unsafe { mem::transmute(self) }
+    }
+}
+
+impl<T> Deref for SlimSliceBox<T> {
+    type Target = [T];
+
+    #[allow(clippy::needless_lifetimes)]
+    fn deref<'a>(&'a self) -> &'a Self::Target {
+        // SAFETY: `ptr` and `len` are derived from a live `Box<[T]>` valid for `'self`
+        // so we satisfy all safety requirements for `from_raw_parts`.
+        unsafe { self.raw.deref() }
+    }
+}
+
+impl<T> DerefMut for SlimSliceBox<T> {
+    #[allow(clippy::needless_lifetimes)]
+    fn deref_mut<'a>(&'a mut self) -> &'a mut Self::Target {
+        // SAFETY: `ptr` and `len` are derived from a live `Box<[T]>`
+        // so we satisfy all safety requirements for `from_raw_parts`.
+        unsafe { self.raw.deref_mut() }
+    }
+}
+
+impl<T> TryFrom<Box<[T]>> for SlimSliceBox<T> {
+    type Error = LenTooLong<Box<[T]>>;
+
+    fn try_from(boxed: Box<[T]>) -> Result<Self, Self::Error> {
+        ensure_len_fits!(boxed);
+        // SAFETY: Checked above that `len <= u32::MAX`.
+        Ok(unsafe { Self::from_boxed_unchecked(boxed) })
+    }
+}
+
+impl<T> TryFrom<Vec<T>> for SlimSliceBox<T> {
```

Comment:
Are we going to end up wanting a slimvec type that sits under satsvec, to interop with all this? Sorry if this is made clear elsewhere, this has been a big patch.